### ISSUE TITLE
Battlefield Hardline

### ISSUE THREAD

**Issue by libcg** [libcg/grvk#39]
Meta-issue for Battlefield Hardline.

- [ ] Working?
- [ ] ???

**Comment by bazookaben** [comment]
Does not work.

Get this error message when trying to start a match:

> 
> Microsoft Visual C++ Runtime Library
> 
> Assertion failed!
> 
> Program: D:\Origin Games\BFH\bfh.exe
> File: ../src/amdilc/amdilc_compiler.c
> Line: 1979
> 
> Expression: false
> 
> For information on how your program can cause an assertion
> failure, see the Visual C++ documentation on asserts
> 
> Press Retry to debug the application - JIT must be enabled


[grvk.log](https://github.com/libcg/grvk/files/7697665/grvk.log)

**Comment by Cherser-s** [comment]
Can you dump the shader binaries?

**Comment by bazookaben** [comment]
How would I do that?

**Comment by niobium93** [comment]
Set the `GRVK_DUMP_SHADERS=1` environment variable.

**Comment by Cherser-s** [comment]
Yup, then the game will probably dump it in "current" directory (which will be probably the game directory)

**Comment by bazookaben** [comment]
OK I brought up the Environment Variable window in Windows, so what do I put in "variable" and what do I put in "value"? Or am I in the wrong place

**Comment by Cherser-s** [comment]
well, in the snippet above, "1" is the value, while "GRVK_DUMP_SHADERS" is the key/name (or "variable").

**Comment by Cherser-s** [comment]
Actually, if you are using Steam, you can set an environment variable there as in `GRVK_DUMP_SHADERS=1 %command%`.

**Comment by bazookaben** [comment]
Ok, I tried the setting that environment variable both in the user and system categories, and in the "launch options" in the Original launcher. No new files appeared in the game directory

**Comment by Cherser-s** [comment]
idk about windows, maybe you need to reboot your pc and check that env var is set through echoing inside cmd.exe. You can set `GRVK_LOG_LEVEL` variable to `debug`, so you get more data in logs (will be easier to find failing shaders). Also download newer grvk libraries, since I've implemented one of the instructions that are being used by the game and weren't implemented (maybe this will compile the shader correctly when you are joining the match, though tess shaders are still not implemented in the moment).

**Comment by niobium93** [comment]
First [open PowerShell in the games folder](https://adamtheautomator.com/windows-open-powershell-in-a-folder/), then you can set variables easily in PowerShell:
```
$env:GRVK_DUMP_SHADERS = 1
```
Then launch the game from the same PowerShell session.
Source:
[https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.2](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.2)
[https://www.tutorialspoint.com/how-to-set-environment-variables-using-powershell](https://www.tutorialspoint.com/how-to-set-environment-variables-using-powershell)

**Comment by bazookaben** [comment]
Yeah restarting my PC was the key

I used build #39, which didn’t crash but just basically ran as a background process I could see in task manager

GitHub won’t let me post a Zip for some reason, so here’s a drive link, includes the log : https://drive.google.com/file/d/1ugKijTgFRIan3xxf-FvaLpIF8WdkcqjC/view?usp=drivesdk

**Comment by libcg** [comment]
Unfortunately I can't test the game because Battlelog doesn't work on Linux...

**Comment by Cherser-s** [comment]
Ok, looks like your shaders get compiled properly with the exception of tessellation shaders, which aren't implemented.
try compiling my branch: https://github.com/Cherser-s/grvk/tree/tessellation-new

**Comment by bazookaben** [comment]
Got an error when I tried to compile:
[meson-log.txt](https://github.com/libcg/grvk/files/8759028/meson-log.txt)

Followed the instructions on the main page. pretty sure I installed all the necessary stuff on my Ubuntu laptop.

**Comment by Cherser-s** [comment]
Hmmmm it looks like it doesn't like the compiler you have installed
considering how init tests fail

**Comment by bazookaben** [comment]
Do you use something other than meson?

**Comment by Cherser-s** [comment]
Nope,
why would it compare `sizeof(void*)` to 15?? it isn't 128bit. Are you sure, that you configured it correctly?

**Comment by niobium93** [comment]
I just tried Battlefield Hardline and it runs fine. As long as Fullscreen Mode is set to Borderless and Terrain Quality to Medium. Here's the log:
[grvk.log](https://github.com/libcg/grvk/files/9094126/grvk.log)

**Comment by niobium93** [comment]
Here's a screenshot:
![2022-07-12-17:49:44-screenshot](https://user-images.githubusercontent.com/1523948/178519124-e31ae7b2-b809-43ed-88b3-625f1d1e47f5.png)

**Comment by niobium93** [comment]
This same scene with the same settings on DXVK reaches 130-150 fps, so GRVK is a lot slower.

**Comment by niobium93** [comment]
@Cherser-s I tried your `tessellation-new` branch and it indeed lets me run the game with max settings.
Log: [grvk.log](https://github.com/libcg/grvk/files/9094462/grvk.log)
Screenshot:
![2022-07-12-18:19:15-screenshot](https://user-images.githubusercontent.com/1523948/178526220-eae3436a-aba1-4d8f-b776-e8bd928c9515.png)
It runs exactly as fast as regular GRVK and I can't really see any difference visually. Oh and the game always crashes when quitting. On either branch.

**Comment by niobium93** [comment]
Do you guys need a shader dump or something before I uninstall this game?

**Comment by Cherser-s** [comment]
> Do you guys need a shader dump
yes, it would be great

> I can't really see any difference visually.
not a surprise, considering the game only uses the tessellation for terrain calculation

idk if there is a perf graph in this game, but can you also check the CPU overhead between grvk and dxvk?

**Comment by niobium93** [comment]
@Cherser-s [Here are the shaders](https://mega.nz/file/3NBnESYa#ePZA10XTFAUcm4h3Az239GGnAiFlC-Tn6XkoBwj4xms)

Log files with `GRVK_LOG_LEVEL=trace`:
[grvk_axl.log.zst](https://mega.nz/file/SYpEGLBR#tbtdCoQ_jFvTIIrUTgNdeDqeF0bi6K0E0lmG5_WWUsE)
[grvk.log.zst](https://mega.nz/file/mJwBRZRD#ZTZQYnXVGmf6j-eE6cp_N3IUZyGvInOkID0bLIPDork)

Is this enough info about the CPU overhead difference?:
![comparison](https://user-images.githubusercontent.com/1523948/178550905-b3d70113-7ede-485f-bb72-7c85a433c3fe.png)

**Comment by Cherser-s** [comment]
No, it just shows the frametime, unfortunately mangohud can't show this.
Open the in-game console and try the commands like `PerfOverlay.DrawGraph 1` (idk how it looks in hardline, I've just written here the one from bf4).

**Comment by niobium93** [comment]
@Cherser-s is that enough?

**Comment by Cherser-s** [comment]
yes, thank you

**Comment by niobium93** [comment]
You're welcome :smile:
